utils/man/man1/powercap-info.1
.TH "POWERCAP-INFO" "1" "2021-12-12" "powercap" "powercap\-info"
.SH "NAME"
.LP
powercap\-info \- get information from the Linux power capping framework
.SH "SYNPOSIS"
.LP
\fBpowercap\-info\fP [\fINAME\fP [\fIOPTION\fP]...] [\fB\-hv\fP]
.SH "DESCRIPTION"
.LP
Prints configurations for powercap control types.
.LP
Output can be filtered by specifying a control type \fINAME\fP and
\fIOPTION\fP flags.
.LP
A control type \fINAME\fP must not be empty or contain a '.' or '/'.
.SH "OPTIONS"
.LP
.TP
\fB\-h,\fR \fB\-\-help\fR
Prints the help screen
.TP
\fB\-v,\fR \fB\-\-verbose\fR
Print errors when files cannot be read
.TP
\fB\-p,\fR \fB\-\-control\-type\fR=\fINAME\fP
Deprecated, provide \fINAME\fP as the first positional argument instead.
.LP
All remaining options below require a control type \fINAME\fP:
.TP
\fB\-z,\fR \fB\-\-zone=\fR\fIZONE(S)\fP
The zone/subzone numbers in the control type's powercap tree (control
type's root by default).
Separate zones/subzones with a colon.
E.g., for zone 0, subzone 2:
.br
\fB\-z 0:2\fP
.br
Ending with a colon prevents output for subzones.
E.g., for zone 0, but not subzones:
.br
\fB\-z 0:\fP
.TP
\fB\-c,\fR \fB\-\-constraint\fR=\fICONSTRAINT\fP
The constraint number
.TP
All remaining options below are mutually exclusive:
.TP
\fB\-E,\fR \fB\-\-enabled\fR
Print control type enabled/disabled status
.TP
\fB\-n,\fR \fB\-\-nzones\fR
Print the number of zones (control type's root by default; within the
\-z/\-\-zone level, if set)
.LP
The following are zone-level arguments and require \-z/\-\-zone:
.TP
\fB\-N,\fR \fB\-\-nconstraints\fR
Print the number of zone constraints
.TP
\fB\-j,\fR \fB\-\-z\-energy\fR
Print zone energy counter
.TP
\fB\-J,\fR \fB\-\-z\-max\-energy\-range\fR
Print zone maximum energy counter range
.TP
\fB\-w,\fR \fB\-\-z\-power\fR
Print zone current power
.TP
\fB\-W,\fR \fB\-\-z\-max\-power\-range\fR
Print zone maximum current power range
.TP
\fB\-e,\fR \fB\-\-z\-enabled\fR
Print zone enabled/disabled status
.TP
\fB\-x,\fR \fB\-\-z\-name\fR
Print zone name
.LP
The following are constraint-level arguments and require \-z/\-\-zone and
\-c/\-\-constraint:
.TP
\fB\-l,\fR \fB\-\-c\-power\-limit\fR
Print constraint power limit
.TP
\fB\-s,\fR \fB\-\-c\-time\-window\fR
Print constraint time window
.TP
\fB\-U,\fR \fB\-\-c\-max\-power\fR
Print constraint maximum allowed power
.TP
\fB\-u,\fR \fB\-\-c\-min\-power\fR
Print constraint minimum allowed power
.TP
\fB\-T,\fR \fB\-\-c\-max\-time\-window\fR
Print constraint maximum allowed time window
.TP
\fB\-t,\fR \fB\-\-c\-min\-time\-window\fR
Print constraint minimum allowed time window
.TP
\fB\-y,\fR \fB\-\-c\-name\fR
Print constraint name
.SH "EXAMPLES"
.TP
\fBpowercap\-info\fP
Print all zones for all control types.
.TP
\fBpowercap\-info intel\-rapl\fP
Print all zones for the \fIintel\-rapl\fR control type.
.TP
\fBpowercap\-info intel\-rapl \-z 0\fP
Print only zone 0 for the \fIintel\-rapl\fR control type.
.TP
\fBpowercap\-info intel\-rapl \-z 0:1\fP
Print zone 0, subzone 1 for the \fIintel\-rapl\fR control type.
.TP
\fBpowercap\-info intel\-rapl \-z 0 \-c 1\fP
Print zone 0, constraint 1 for the \fIintel\-rapl\fR control type.
.TP
\fBpowercap\-info intel\-rapl \-z 0 \-j\fP
Print the energy counter for zone 0 for the \fIintel\-rapl\fR control type.
.TP
\fBpowercap\-info intel\-rapl \-z 1:0 \-c 0 \-l\fP
Print the power limit for zone 1, subzone 0, constraint 0 for the
\fIintel\-rapl\fR control type.
.SH "REMARKS"
.LP
Some fields are optional and/or may require administrative (super\-user)
privileges to read.
.br
Fields will only be printed if they are available and readable, unless
\-v/\-\-verbose is set.
.br
If no subzone/constraint\-specific outputs are requested, all available
zones and constraints will be shown.
.LP
Energy units: microjoules (uJ)
.br
Power units: microwatts (uW)
.br
Time units: microseconds (us)
.SH "BUGS"
.LP
Report bugs upstream at <https://github.com/powercap/powercap>
.SH "FILES"
.nf
\fI/sys/devices/virtual/powercap/*\fP
.nf
\fI/sys/class/powercap/*\fP
.fi
.SH "AUTHORS"
.nf
Connor Imes <connor.k.imes@gmail.com>
.fi
.SH "SEE ALSO"
.BR powercap\-set (1)

RELEASES.md
# Release Notes

## [Unreleased]

### Added

* powercap-info: control type name no longer required - will print info for all control types if not set


## [v0.5.0] - 2021-12-09

### Added

* CMake installation components

### Changed

* Increased minimum CMake version from 3.1 to 3.12
* CMake package config now exports binaries in a 'Utils' component rather than parent package


## [v0.4.0] - 2021-04-25

### Fixed

* powercap.h: header documentation for 'powercap_*_file_get_name' functions' return behavior
* rapl-{info,set}: deprecation messages that specified non-existent alternatives

### Added

* Functions to support opening files in powercap.h:
  * 'powercap_control_type_file_open'
  * 'powercap_zone_file_open'
  * 'powercap_constraint_file_open'
* CMake helper to find powercap without relying on pkg-config

### Changed

* Binaries powercap-{info,set} now accept the control type argument without requiring a preceding -p flag
* Increased minimum CMake version from 2.8.12 to 3.1

### Deprecated

* Flag '-p/--control-type' in powercap-{info,set} - specify control type name as the first positional argument instead
* Add compiler deprecation attributes to deprecated functions


## [v0.3.1] - 2020-11-08

### Fixed

* [#6] powercap-common-test:test_snprintf_base_path fails on ppc64el architecture


## [v0.3.0] - 2020-08-15

### Added

* Support for top-level control type
  * Struct 'powercap_control_type', enum 'powercap_control_type_file', and associated functions in powercap.h:
    * 'powercap_control_type_get_enabled'
    * 'powercap_control_type_set_enabled'
    * 'powercap_control_type_file_get_name'
  * Functions in powercap-sysfs.h:
    * 'powercap_sysfs_control_type_get_enabled'
    * 'powercap_sysfs_control_type_set_enabled'
  * Functions in powercap-rapl.h:
    * 'powercap_rapl_control_is_supported'
    * 'powercap_rapl_control_is_enabled'
    * 'powercap_rapl_control_set_enabled'
  * Argument '--enabled' to powercap-{info,set} for getting/setting control type enabled field
* Function 'powercap_rapl_get_num_instances' in powercap-rapl.h (supersedes 'powercap_rapl_get_num_packages')
* Argument '--nconstraints' to powercap-info for getting the number of zone constraints

### Changed

* Increased minimum CMake version from 2.8.5 to 2.8.12 to support target_compile_definitions
* Updated 'powercap-{info,set}' man pages
* Argument '--zone' for powercap-set no longer required due to introduction of '--enabled' argument
* Disabled logging by default (in general, libraries shouldn't print output unless requested)

### Deprecated

* Binaries rapl-{info,set} - use powercap-{info,set} instead
* Interface powercap-rapl-sysfs.h - use powercap-sysfs.h directly instead
* Function 'powercap_rapl_get_num_packages' in powercap-rapl.h - use 'powercap_rapl_get_num_instances' instead

### Fixed

* Fixed [#4]: powercap-rapl now checks if parent zone is PACKAGE or PSYS


## [v0.2.0] - 2019-12-03

### Added

* This RELEASES.md file
* Multiarch support (use GNU standard installation directories)
* Additional documentation in README
* Function 'rapl_sysfs_zone_exists' in powercap-rapl-sysfs.h
* Command line long option '--zone' to rapl-info and rapl-set
* Functions to powercap-rapl-sysfs.h allowed by the powercap API that RAPL doesn't support (but could in the future)
* Options to rapl-info and rapl-set to support optional powercap features not currently implemented in RAPL

### Changed

* Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
* More pedantic man page source formatting
* RAPL sysfs 'pkg' parameter names changed to 'zone' to improve generality
* RAPL 'package' parameter names changed to 'id' to improve generality
* On failure, binaries now exit with positive error codes instead of negative values

### Deprecated

* Function 'rapl_sysfs_pkg_exists' in powercap-rapl-sysfs.h - use 'rapl_sysfs_zone_exists' instead
* Function 'rapl_sysfs_sz_exists' in powercap-rapl-sysfs.h - use 'rapl_sysfs_zone_exists' instead
* Command line long option '--package' for rapl-info and rapl-set - use '--zone' instead

### Removed

* Removed private symbol exports in shared object library (already patched in Debian)
* Undocumented (and unintended) --package long option from powercap-set

### Fixed

* Fixed [#1]: Kernel uses hexadecimal numbers in directory paths, not decimal


## [v0.1.1] - 2017-09-21

### Added

* Added stateless interfaces to sysfs
* Added powercap-info, powercap-set, rapl-info, and rapl-set binaries and man pages
* Added VERSION and SOVERSION to shared object libraries

### Changed

* Update license to use author as copyright holder


## v0.1.0 - 2017-06-09

### Added

* Initial public release

[Unreleased]: https://github.com/powercap/powercap/compare/v0.5.0...HEAD
[v0.5.0]: https://github.com/powercap/powercap/compare/v0.4.0...v0.5.0
[v0.4.0]: https://github.com/powercap/powercap/compare/v0.3.1...v0.4.0

utils/powercap-info.c
/*
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Get powercap values.
 *
 * @author Connor Imes
 * @date 2017-08-24
 */
#include <dirent.h>
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include "powercap-sysfs.h"
#include "util-common.h"

#define POWERCAP_PATH "/sys/class/powercap"

static void print_parent_headers(const uint32_t* zones, uint32_t depth_start, uint32_t depth, uint32_t indnt) {
  uint32_t i;
  uint32_t j;
  for (i = depth_start; i <= depth; i++) {
    indent(indnt + i - 1);
    printf("Zone %"PRIu32, zones[0]);
    for (j = 1; j < i; j++) {
      printf(":%"PRIu32, zones[j]);
    }
    printf("\n");
  }
}

static void analyze_constraint(const char* control_type, const uint32_t* zones, uint32_t depth, uint32_t constraint, int verbose, uint32_t indnt) {
  char name[MAX_NAME_SIZE];
  uint64_t val64;
  ssize_t sret;
  int ret;

  indent(indnt + depth);
  printf("Constraint %"PRIu32"\n", constraint);

  sret = powercap_sysfs_constraint_get_name(control_type, zones, depth, constraint, name, sizeof(name));
  ret = sret > 0 ? 0 : (int) sret;
  str_or_verbose(verbose, indnt + depth + 1, "name", name, ret);

  ret = powercap_sysfs_constraint_get_power_limit_uw(control_type, zones, depth, constraint, &val64);
  u64_or_verbose(verbose, indnt + depth + 1, "power_limit_uw", val64, ret);

  ret = powercap_sysfs_constraint_get_time_window_us(control_type, zones, depth, constraint, &val64);
  u64_or_verbose(verbose, indnt + depth + 1, "time_window_us", val64, ret);

  ret = powercap_sysfs_constraint_get_min_power_uw(control_type, zones, depth, constraint, &val64);
  u64_or_verbose(verbose, indnt + depth + 1, "min_power_uw", val64, ret);

  ret = powercap_sysfs_constraint_get_max_power_uw(control_type, zones, depth, constraint, &val64);
  u64_or_verbose(verbose, indnt + depth + 1, "max_power_uw", val64, ret);

  ret = powercap_sysfs_constraint_get_min_time_window_us(control_type, zones, depth, constraint, &val64);
  u64_or_verbose(verbose, indnt + depth + 1, "min_time_window_us", val64, ret);

  ret = powercap_sysfs_constraint_get_max_time_window_us(control_type, zones, depth, constraint, &val64);
  u64_or_verbose(verbose, indnt + depth + 1, "max_time_window_us", val64, ret);
}

static void analyze_zone(const char* control_type, const uint32_t* zones, uint32_t depth, int verbose, uint32_t indnt) {
  char name[MAX_NAME_SIZE];
  uint64_t val64;
  uint32_t val32;
  ssize_t sret;
  int ret;

  print_parent_headers(zones, depth, depth, indnt);

  sret = powercap_sysfs_zone_get_name(control_type, zones, depth, name, sizeof(name));
  ret = sret > 0 ? 0 : (int) sret;
  str_or_verbose(verbose, indnt + depth, "name", name, ret);

  ret = powercap_sysfs_zone_get_enabled(control_type, zones, depth, &val32);
  u64_or_verbose(verbose, indnt + depth, "enabled", (uint64_t) val32, ret);

  ret = powercap_sysfs_zone_get_max_energy_range_uj(control_type, zones, depth, &val64);
  u64_or_verbose(verbose, indnt + depth, "max_energy_range_uj", val64, ret);

  ret = powercap_sysfs_zone_get_energy_uj(control_type, zones, depth, &val64);
  u64_or_verbose(verbose, indnt + depth, "energy_uj", val64, ret);

  ret = powercap_sysfs_zone_get_max_power_range_uw(control_type, zones, depth, &val64);
  u64_or_verbose(verbose, indnt + depth, "max_power_range_uw", val64, ret);

  ret = powercap_sysfs_zone_get_power_uw(control_type, zones, depth, &val64);
  u64_or_verbose(verbose, indnt + depth, "power_uw", val64, ret);

  for (val32 = 0; !powercap_sysfs_constraint_exists(control_type, zones, depth, val32); val32++) {
    analyze_constraint(control_type, zones, depth, val32, verbose, indnt);
  }
}

static void analyze_control_type(const char* control_type, int verbose, uint32_t indnt) {
  uint32_t val32;
  int ret = powercap_sysfs_control_type_get_enabled(control_type, &val32);
  u64_or_verbose(verbose, indnt, "enabled", (uint64_t) val32, ret);
}

/* depth must be > 0 */
static void analyze_all_zones_recurse(const char* control_type, uint32_t* zones, uint32_t depth, uint32_t max_depth, int verbose, uint32_t indnt) {
  if (!powercap_sysfs_zone_exists(control_type, zones, depth)) {
    /* Analyze this zone */
    analyze_zone(control_type, zones, depth, verbose, indnt);
    if (depth < max_depth) {
      /* Analyze subzones */
      zones[depth] = 0;
      analyze_all_zones_recurse(control_type, zones, depth + 1, max_depth, verbose, indnt);
    }
    /* Analyze next sibling zone */
    zones[depth - 1]++;
    analyze_all_zones_recurse(control_type, zones, depth, max_depth, verbose, indnt);
  }
}

static void analyze_control_type_recurse(const char* control_type, uint32_t* zones, uint32_t max_depth, int verbose, uint32_t indnt) {
  analyze_control_type(control_type, verbose, indnt);
  analyze_all_zones_recurse(control_type, zones, 1, max_depth, verbose, indnt);
}

static int analyze_powercap(uint32_t* zones, uint32_t max_depth, int verbose) {
  struct dirent *dp;
  DIR *dfd;
  if ((dfd = opendir(POWERCAP_PATH)) == NULL) {
    perror(POWERCAP_PATH);
    return -errno;
  }
  while ((dp = readdir(dfd)) != NULL) {
    if (dp->d_name && !strstr(dp->d_name, ".") && !strstr(dp->d_name, ":")) {
      memset(zones, 0, max_depth * sizeof(*zones));
      printf("%s\n", dp->d_name);
      analyze_control_type_recurse(dp->d_name, zones, max_depth, verbose, 1);
    }
  }
  if (closedir(dfd)) {
    perror(POWERCAP_PATH);
    return -errno;
  }
  return 0;
}

static void analyze_zone_recurse(const char* control_type, uint32_t* zones, uint32_t depth, uint32_t max_depth, int verbose, uint32_t indnt) {
  if (!powercap_sysfs_zone_exists(control_type, zones, depth)) {
    /* Analyze this zone */
    analyze_zone(control_type, zones, depth, verbose, indnt);
    if (depth < max_depth) {
      /* Analyze subzones */
      zones[depth] = 0;
      analyze_all_zones_recurse(control_type, zones, depth + 1, max_depth, verbose, indnt);
    }
  }
}

static void print_num_zones(const char* control_type, uint32_t* zones, uint32_t depth) {
  zones[depth] = 0;
  while (!powercap_sysfs_zone_exists(control_type, zones, depth + 1)) {
    zones[depth]++;
  }
  printf("%"PRIu32"\n", zones[depth]);
}

static void print_num_constraints(const char* control_type, uint32_t* zones, uint32_t depth) {
  uint32_t n = 0;
  while (!powercap_sysfs_constraint_exists(control_type, zones, depth, n)) {
    n++;
  }
  printf("%"PRIu32"\n", n);
}

static const char short_options[] = "-hvp:z:c:EnNjJwWexlsUuTty";

static const struct option long_options[] = {
  {"help",                no_argument,        NULL, 'h'},
  {"verbose",             no_argument,        NULL, 'v'},
  {"control-type",        required_argument,  NULL, 'p'},
  {"zone",                required_argument,  NULL, 'z'},
  {"constraint",          required_argument,  NULL, 'c'},
  {"enabled",             no_argument,        NULL, 'E'},
  {"nzones",              no_argument,        NULL, 'n'},
  {"nconstraints",        no_argument,        NULL, 'N'},
  {"z-energy",            no_argument,        NULL, 'j'},
  {"z-max-energy-range",  no_argument,        NULL, 'J'},
  {"z-power",             no_argument,        NULL, 'w'},
  {"z-max-power-range",   no_argument,        NULL, 'W'},
  {"z-enabled",           no_argument,        NULL, 'e'},
  {"z-name",              no_argument,        NULL, 'x'},
  {"c-power-limit",       no_argument,        NULL, 'l'},
  {"c-time-window",       no_argument,        NULL, 's'},
  {"c-max-power",         no_argument,        NULL, 'U'},
  {"c-min-power",         no_argument,        NULL, 'u'},
  {"c-max-time-window",   no_argument,        NULL, 'T'},
  {"c-min-time-window",   no_argument,        NULL, 't'},
  {"c-name",              no_argument,        NULL, 'y'},
  {0, 0, 0, 0}
};

static void print_usage(void) {
  printf("Usage: powercap-info [NAME [OPTION]...] [-hv]\n\n");
  printf("Prints configurations for powercap control types.\n");
  printf("Output can be filtered by specifying a control type NAME and OPTION flags.\n");
  printf("A control type NAME must not be empty or contain a '.' or '/'.\n\n");
  printf("Options:\n");
  printf("  -h, --help                   Print this message and exit\n");
  printf("  -v, --verbose                Print errors when files cannot be read\n");
  printf("  -p, --control-type=NAME      Deprecated, provide NAME as the first\n");
  printf("                               positional argument instead\n");
  printf("All remaining options below require a control type NAME:\n");
  printf("  -z, --zone=ZONE(S)           The zone/subzone numbers in the control type's\n");
  printf("                               powercap tree (control type's root by default)\n");
  printf("                               Separate zones/subzones with a colon\n");
  printf("                               E.g., for zone 0, subzone 2: \"-z 0:2\"\n");
  printf("                               Ending with a colon prevents output for subzones\n");
  printf("                               E.g., for zone 0, but not subzones: \"-z 0:\"\n");
  printf("  -c, --constraint=CONSTRAINT  The constraint number\n");
  printf("All remaining options below are mutually exclusive:\n");
  printf("  -E, --enabled                Print control type enabled/disabled status\n");
  printf("  -n, --nzones                 Print the number of zones (control type's root by\n");
  printf("                               default; within the -z/--zone level, if set)\n");
  printf("The following are zone-level arguments and require -z/--zone:\n");
  printf("  -N, --nconstraints           Print the number of zone constraints\n");
  printf("  -j, --z-energy               Print zone energy counter\n");
  printf("  -J, --z-max-energy-range     Print zone maximum energy counter range\n");
  printf("  -w, --z-power                Print zone current power\n");
  printf("  -W, --z-max-power-range      Print zone maximum current power range\n");
  printf("  -e, --z-enabled              Print zone enabled/disabled status\n");
  printf("  -x, --z-name                 Print zone name\n");
  printf("The following are constraint-level arguments and require -z/--zone and -c/--constraint:\n");
  printf("  -l, --c-power-limit          Print constraint power limit\n");
  printf("  -s, --c-time-window          Print constraint time window\n");
  printf("  -U, --c-max-power            Print constraint maximum allowed power\n");
  printf("  -u, --c-min-power            Print constraint minimum allowed power\n");
  printf("  -T, --c-max-time-window      Print constraint maximum allowed time window\n");
  printf("  -t, --c-min-time-window      Print constraint minimum allowed time window\n");
  printf("  -y, --c-name                 Print constraint name\n");
  printf("\nSome fields are optional and/or may require administrative (super-user) privileges to read.\n");
  printf("Fields will only be printed if they are available and readable, unless -v/--verbose is set.\n");
  printf("If no zone/constraint-specific outputs are requested, all available zones and constraints will be shown.\n");
  printf("\nEnergy units: microjoules (uJ)\n");
  printf("Power units: microwatts (uW)\n");
  printf("Time units: microseconds (us)\n");
}

static void print_common_help(void) {
  printf("Considerations for common errors:\n");
  printf("- Ensure that the control type exists, which may require loading a kernel module\n");
  printf("- Some files may simply not exist\n");
}

static int verify_control_type_args(const char* control_type, uint32_t depth, u32_param* constraint, int unique_set) {
  int ret = 0;
  if (!is_valid_powercap_control_type(control_type)) {
    fprintf(stderr, "Must specify control type NAME; value must not be empty or contain any '.' or '/' characters\n");
    ret = -EINVAL;
  } else if (unique_set) {
    switch (unique_set) {
    case 'E':
      if (depth || constraint->set) {
        fprintf(stderr, "Must not specify -z/--zone or -c/--constraint with -E/--enabled\n");
        ret = -EINVAL;
      }
      break;
    case 'n':
      if (constraint->set) {
        fprintf(stderr, "Must not specify -c/--constraint with -n/--nzones\n");
        ret = -EINVAL;
      }
      break;
    case 'N':
    case 'j':
    case 'J':
    case 'w':
    case 'W':
    case 'e':
    case 'x':
      if (!depth) {
        fprintf(stderr, "Must specify -z/--zone with zone-level argument\n");
        ret = -EINVAL;
      } else if (constraint->set) {
        fprintf(stderr, "Must not specify -c/--constraint with zone-level argument\n");
        ret = -EINVAL;
      }
      break;
    case 'l':
    case 's':
    case 'U':
    case 'u':
    case 'T':
    case 't':
    case 'y':
      if (!depth || !constraint->set) {
        fprintf(stderr, "Must specify -z/--zone and -c/--constraint with constraint-level argument\n");
        ret = -EINVAL;
      }
      break;
    }
  } else if (constraint->set && !depth) {
    fprintf(stderr, "Must specify -z/--zone with -c/--constraint\n");
    ret = -EINVAL;
  }
  return ret;
}

static int print_control_type(const char* control_type, uint32_t* zones, uint32_t depth, uint32_t max_depth,
                              u32_param* constraint, int recurse, int verbose, int unique_set, uint32_t indnt) {
  uint64_t val64;
  uint32_t val32;
  char name[MAX_NAME_SIZE];
  int ret = 0;

  /* Check if control type/zones/constraint exist */
  if (powercap_sysfs_control_type_exists(control_type)) {
    fprintf(stderr, "Control type does not exist\n");
    return -EINVAL;
  } else if (depth && powercap_sysfs_zone_exists(control_type, zones, depth)) {
    fprintf(stderr, "Zone does not exist\n");
    return -EINVAL;
  } else if (constraint->set && powercap_sysfs_constraint_exists(control_type, zones, depth, constraint->val)) {
    fprintf(stderr, "Constraint does not exist\n");
    return -EINVAL;
  }

  /* Perform requested action */
  if (unique_set) {
    switch (unique_set) {
    case 'E':
      /* Get control type enabled */
      if (!(ret = powercap_sysfs_control_type_get_enabled(control_type, &val32))) {
        printf("%"PRIu32"\n", val32);
      } else {
        perror("Failed to get control type enabled");
      }
      break;
    case 'n':
      /* Print number of zones at the specified tree location */
      print_num_zones(control_type, zones, depth);
      break;
    case 'N':
      /* Print number of constraints at the specified tree location */
      print_num_constraints(control_type, zones, depth);
      break;
    case 'j':
      /* Get zone energy */
      if (!(ret = powercap_sysfs_zone_get_energy_uj(control_type, zones, depth, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get zone energy");
      }
      break;
    case 'J':
      /* Get zone max energy range */
      if (!(ret = powercap_sysfs_zone_get_max_energy_range_uj(control_type, zones, depth, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get zone max energy range");
      }
      break;
    case 'w':
      /* Get zone power */
      if (!(ret = powercap_sysfs_zone_get_power_uw(control_type, zones, depth, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get zone power");
      }
      break;
    case 'W':
      /* Get zone max power range */
      if (!(ret = powercap_sysfs_zone_get_max_power_range_uw(control_type, zones, depth, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get zone max power range");
      }
      break;
    case 'e':
      /* Get zone enabled */
      if (!(ret = powercap_sysfs_zone_get_enabled(control_type, zones, depth, &val32))) {
        printf("%"PRIu32"\n", val32);
      } else {
        perror("Failed to get zone enabled");
      }
      break;
    case 'x':
      /* Get zone name */
      if (powercap_sysfs_zone_get_name(control_type, zones, depth, name, sizeof(name)) > 0) {
        printf("%s\n", name);
      } else {
        ret = -errno;
        perror("Failed to get zone name");
      }
      break;
    case 'l':
      /* Get constraint power limit */
      if (!(ret = powercap_sysfs_constraint_get_power_limit_uw(control_type, zones, depth, constraint->val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint power limit");
      }
      break;
    case 's':
      /* Get constraint time window */
      if (!(ret = powercap_sysfs_constraint_get_time_window_us(control_type, zones, depth, constraint->val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint time window");
      }
      break;
    case 'U':
      /* Get constraint max power */
      if (!(ret = powercap_sysfs_constraint_get_max_power_uw(control_type, zones, depth, constraint->val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint max power");
      }
      break;
    case 'u':
      /* Get constraint min power */
      if (!(ret = powercap_sysfs_constraint_get_min_power_uw(control_type, zones, depth, constraint->val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint min power");
      }
      break;
    case 'T':
      /* Get constraint max time window */
      if (!(ret = powercap_sysfs_constraint_get_max_time_window_us(control_type, zones, depth, constraint->val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint max time window");
      }
      break;
    case 't':
      /* Get constraint min time window */
      if (!(ret = powercap_sysfs_constraint_get_min_time_window_us(control_type, zones, depth, constraint->val, &val64))) {
        printf("%"PRIu64"\n", val64);
      } else {
        perror("Failed to get constraint min time window");
      }
      break;
    case 'y':
      /* Get constraint name */
      if (powercap_sysfs_constraint_get_name(control_type, zones, depth, constraint->val, name, sizeof(name)) > 0) {
        printf("%s\n", name);
      } else {
        ret = -errno;
        perror("Failed to get constraint name");
      }
      break;
    }
  } else if (depth > 0) {
    /* Print summary of zone or constraint */
    if (constraint->set) {
      /* print constraint */
      print_parent_headers(zones, 1, depth, 0);
      analyze_constraint(control_type, zones, depth, constraint->val, verbose, 0);
    } else {
      /* print zone */
      print_parent_headers(zones, 1, depth - 1, 0);
      if (recurse) {
        analyze_zone_recurse(control_type, zones, depth, max_depth, verbose, 0);
      } else {
        analyze_zone(control_type, zones, depth, verbose, 0);
      }
    }
  } else {
    /* print control type and all zones */
    analyze_control_type_recurse(control_type, zones, max_depth, verbose, 0);
  }
  return ret;
}

int main(int argc, char** argv) {
  const char* control_type = NULL;
  uint32_t zones[MAX_ZONE_DEPTH] = { 0 };
  u32_param constraint = {0, 0};
  uint32_t depth = 0;
  int recurse = 1;
  int verbose = 0;
  int unique_set = 0;
  int c;
  int cont = 1;
  int ret = 0;

  /* Parse command-line arguments */
  while (cont) {
    c = getopt_long(argc, argv, short_options, long_options, NULL);
    switch (c) {
    case -1:
      cont = 0;
      break;
    case 'h':
      print_usage();
      return EXIT_SUCCESS;
    case 'v':
      verbose = 1;
      break;
    case 1:
    case 'p':
      if (control_type) {
        cont = 0;
        ret = -EINVAL;
      }
      control_type = optarg;
      break;
    case 'z':
      recurse = get_recurse(optarg);
      ret = parse_zones(optarg, zones, MAX_ZONE_DEPTH, &depth, &cont);
      break;
    case 'c':
      ret = set_u32_param(&constraint, optarg, &cont);
      break;
    case 'E':
    case 'n':
    case 'N':
    case 'j':
    case 'J':
    case 'w':
    case 'W':
    case 'e':
    case 'x':
    case 'l':
    case 's':
    case 'U':
    case 'u':
    case 'T':
    case 't':
    case 'y':
      if (unique_set) {
        fprintf(stderr, "Must not specify multiple mutually exclusive arguments\n");
        cont = 0;
        ret = -EINVAL;
        break;
      }
      unique_set = c;
      break;
    case '?':
    default:
      cont = 0;
      ret = -EINVAL;
      break;
    }
  }

  /* Verify argument combinations */
  if (ret) {
    fprintf(stderr, "Invalid arguments\n");
  } else if (control_type) {
    ret = verify_control_type_args(control_type, depth, &constraint, unique_set);
  } else if (depth || constraint.set || unique_set) {
    fprintf(stderr, "Must not specify control type options without NAME\n");
    ret = -EINVAL;
  }
  if (ret) {
    print_usage();
    return EXIT_FAILURE;
  }

  /* Print requested info */
  if (control_type) {
    ret = print_control_type(control_type, zones, depth, MAX_ZONE_DEPTH, &constraint, recurse, verbose, unique_set, 0);
  } else {
    ret = analyze_powercap(zones, MAX_ZONE_DEPTH, verbose);
  }
  if (ret) {
    print_common_help();
  }

  return ret ? EXIT_FAILURE : EXIT_SUCCESS;
}
